_posts/2023-06-09-flat-object.md

```diff
@@ -0,0 +1,365 @@
+---
+layout: post
+title:  "Use flat object in OpenSearch"
+authors:
+- mingshl
+- kolchfa
+date: 2023-06-09
+categories:
+ - technical-post
+meta_keywords: 
+meta_description: 
+
+excerpt: OpenSearch 2.7 introduced the new flat object field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The flat object field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the flat object field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch. 
+---
+
+OpenSearch 2.7 introduced the new `flat_object` field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The `flat_object` field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the `flat_object` field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch. 
+
+## Dynamic mapping
+
+In OpenSearch, a [_mapping_](https://opensearch.org/docs/latest/field-types/index/) defines the structure of your data. It specifies field names, types, and indexing and analysis settings, ensuring that your data is organized and interpreted correctly. If you don't specify a custom mapping, OpenSearch infers the structure of your document automatically when you upload the document. This process is called _dynamic mapping_, where OpenSearch detects the document data structure and generates the corresponding mapping file. 
+
+## When dynamic mapping falls flat
+
+When documents have complex data structures or deeply nested fields, relying on dynamic mapping can lead to the number of mapped fields in an index quickly growing to hundreds or even thousands. This "mapping explosion" negatively impacts the performance of your cluster. Additionally, searching through deeply nested indexes with lengthy dot paths can be inconvenient, especially if you are unfamiliar with the document structure. Flat object solves both of these problems.
```

Comment:
can more detail be added about the negative impact of mapping explosions? What actually happens there? Are we able to reproduce a mapping explosion with this use case?

Reply:
the example will be a mapping file that contains a lot of fields. But it maybe too long to add into the blog post. 

How about adding the error messages/symptoms  that you might get when you have mapping explosion issue? 

- OutOfMemoryError： If the mapping becomes too large to fit in memory, you may encounter an OutOfMemoryError, resulting in the unavailability of the cluster or nodes.
- MapperParsingException : If the number of unique field names grows too large, the cluster may throw exceptions such as MapperParsingException or IllegalArgumentException, indicating that the mapping update has failed.
- Performance degradation: As the mapping grows, the performance of indexing and search operations can degrade. Handling a large number of fields requires more resources and processing time, potentially leading to slower indexing and increased query latencies.
- Increased storage requirements: Each field in the mapping needs storage space. With a mapping explosion, the index's storage requirements can significantly increase, impacting disk space utilization and potentially leading to resource constraints.